api/envoy/config/listener/v3/listener_components.proto

```diff
@@ -165,9 +166,25 @@ message FilterChainMatch {
   repeated string application_protocols = 10;
 }
 
+// The configuration for on-demand filter chain. If this field is not empty in FilterChain message,
+// a filter chain will be built on-demand.
+// On-demand filter chains will be first built as placeholders when a listener is created in order to
+// speedup the warming up of the listener. The building and initialization of an on-demand filter chain
+// will be postponed to the arrival of new connections that require this filter chain.
+// The building process of on-demand filter chains will be similar to the one of normal filter chains.
+// However, a unique rebuilder with a unique init manager will be assigned to build each on-demand filter chain.
+// On the completion of building the on-demand filter chain, connections that require this filter chain will be retried.
```

Comment:
What does retried mean? Do you mean the connection will be held until the filter chain build is complete?

Reply:
Yes. The whole process goes like this:
1. master thread build filter chain placeholder.
2. on new connections, filter chain is found to be a placeholder, so workers will post rebuilding request to master thread and the sockets passed to ```onAccept``` will be stored.
3. The rebuilding process runs just like the building of normal filter chains. (The only difference is that now a unique rebuilder will be assigned to manage the life time and context of the filter chain.
4. On completion of the rebuilding, master thread will post callback to worker thread, so that listeners could retry sockets that we previously stored.

Reply:
I understand how it works. :) My point is this is end user documentation. Please write it as such!

Reply:
Yes, thanks matt for reminding me of that, I put too many details that should be transparent to end users.

Reply:
I will update them to more end user-oriented.